Specify Java versions to use (#54)
scalar-labs/scalardl: docs/getting-started.md

ScalarDL (Ledger and Auditor) abstracts data as a set of assets, where each asset is composed of the history of a record identified by a key called `asset_id` and a historical version number called `age`.
In this document, you will create a very simple application to manage an asset's status using ScalarDL Client SDK.

## Install a JDK

> Because ScalarDL is written in Java, using Java is one of the easist ways to build your ScalarDL applications.

Author: @brfrn169 @josh-wong 
I followed the ScalarDB description, but the current one in ScalarDB is a bit misleading since it sounds like we must use Java to use ScalarDB.

Reviewer: I think I understand. 

Basically, a JDK is required only because we're suggesting it as the easiest way to build applications even though there are other languages that we can use to build applications. Is that correct?

If so, I have just a couple minor changes to this sentence:

```suggestion
Because ScalarDL is written in Java, using Java is one of the easiest ways to build a ScalarDL application.
```

Reviewer 2: > I followed the ScalarDB description, but the current one in ScalarDB is a bit misleading since it sounds like we must use Java to use ScalarDB.

Correct, our current documentation specifies that users must use Java:
https://github.com/scalar-labs/scalardb/blob/v3.11.0/docs/getting-started-with-scalardb.md#install-a-jdk

I believe that users are required to use Java when utilizing ScalarDB through the Java library (the community edition). However, Java is not required for using ScalarDB through the ScalarDB Cluster (the enterprise edition). Therefore, if users do not have a license for the enterprise edition, they must use Java.

@feeblefakie @josh-wong 
Does that make sense?

In that case, how about mentioning the ScalarDB Cluster thing (users can use non-Java languages through ScalarDB Cluster) as a note in the document?

In such a case, you must have one of the following Java Development Kits (JDKs) installed in your environment:

- [Oracle JDK](https://www.oracle.com/java/technologies/downloads/) LTS version (8, 11, or 17)
- [OpenJDK](https://openjdk.org/install/) LTS version (8, 11, or 17)

> Also, since ScalarDL is built with JDK 8, you need to use JDK 8 to build your contracts.

Author: ```suggestion
Also, since ScalarDL is built with JDK 8, you must use JDK 8 to build your contracts.
```

> If you use JDK versions other than 8, specify `-release 8` option for `javac` or define what toolchain to use in your Gradle and Maven configurations.

Author: @komamitsu @choplin 
I'm not sure if this is a proper expression. PTAL!

Reviewer 3: Revision proposal:
```
Also, since ScalarDL is built with JDK 8, you must build contracts as a JDK 8-compatible binary.
If you use JDK versions other than 8, you must configure your build tool to build the JDK 8-compatible binary.
There are several ways to specify binary compatibility, including using the `-release 8` option for `javac` or setting Gradle and Maven configurations to use the JDK 8 toolchain.
For example, to specify toolchains in your Gradle and Maven configurations, check the Gradle doc or the Maven doc. 
```

Reviewer 3: I guess it would also be helpful to provide examples of configuration to set toolchains. It would be only a few lines, like [this](https://github.com/scalar-labs/scalardb-analytics-spark/blob/f1accdf0119b2c158c1345ef79a99a88e1dd7661/lib/build.gradle.kts#L26-L30)

Author: @choplin 
Thanks!
Updated the PR based on your feedback. PTAL again!
[33acaf5](https://github.com/scalar-labs/scalardl/pull/54/commits/33acaf58129f284d62598262e8ac9095c7490e27)

Reviewer 3: Looks good! Thank you!

> If you use JDK versions other than 8, specify `-release 8` option for `javac` or define what toolchain to use in your Gradle and Maven configurations.

Reviewer 4: ```suggestion
If you use JDK versions other than 8, specify `--release 8` option for `javac` or define what toolchain to use in your Gradle and Maven configurations.
```
according to the description
```
$ javac -help | grep -A 1 '\--release'
  :
--
  --release <release>
        Compile for a specific VM version. Supported targets: 6, 7, 8, 9, 10, 11
```

Author: Oops, thanks!

Reviewer: My suggestion below includes [@/komamitsu suggested change](https://github.com/scalar-labs/scalardl/pull/54/files#r1483902153).

```suggestion
If you use a version other than JDK 8, specify the `--release 8` option for `javac` or define which toolchain to use in your Gradle and Maven configurations.
```

> If you use JDK versions other than 8, specify `-release 8` option for `javac` or define what toolchain to use in your Gradle and Maven configurations.

Reviewer 5: Can we highlight this part as Warning or Note in the different color,  just like ScalarDB document, so that the customers don't miss it? 
(https://scalardb.scalar-labs.com/docs/latest/getting-started-with-scalardb/)

Author: @NatsukiMiyamoto Yes, we can. But, `Warning` and `Note` are something we use for side information, so I'm not sure if they are proper. 
@josh-wong What would be a good way to emphasize an important description?

Reviewer: I think we should use an `Attention` notice in this case.

My suggestion below includes [@/komamitsu suggested change](https://github.com/scalar-labs/scalardl/pull/54/files#r1483902153).

```suggestion
{% capture notice--warning %}
**Attention**

Since ScalarDL is built on JDK 8, you must use JDK 8 to build your contracts.
If you use a version other than JDK 8, specify the `--release 8` option for `javac` or define which toolchain to use in your Gradle and Maven configurations.
{% endcapture %}

<div class="notice--warning">{{ notice--warning | markdownify }}</div>

```

Reviewer 5: I agree with Josh. Attention would be the right phrase for this case. 
Thank you :)

> For the Gradle and Maven configurations, check [the Gradle doc](https://docs.gradle.org/current/userguide/toolchains.html) and [the Maven doc](https://maven.apache.org/guides/mini/guide-using-toolchains.html).

Reviewer: ```suggestion
For the Gradle and Maven configurations, see [Toolchains for JVM projects for Gradle](https://docs.gradle.org/current/userguide/toolchains.html) and [Guide to Using Toolchains for Maven](https://maven.apache.org/guides/mini/guide-using-toolchains.html).
```

{% capture notice--info %}
**Note**

> We recommend using the LTS versions mentioned above, but other non-LTS versions may work.

Reviewer 6: From the product/technical perspective, I think there is no problem with this description.
However, from the support perspective, I feel `may work` is a bit ambiguous.
Sometimes, I get an inquiry such as `What Java version do you support?` from a customer on the support site.

So, I want to define whether we will support the non-LTS version or not in the support site for users who have a commercial license.

If we don't support the not-LTS version, it might be better to add a note like `We don't provide technical support for non-LTS use cases even if you have a commercial license.`.

If such a description is not appropriate in this document (since this document is a technical document), I want to discuss it in a support team meeting.

@feeblefakie @NatsukiMiyamoto 
What do you think?

Reviewer 5: @kota2and3kan  Let's talk separately. I agree your points that this may be a bit dangerous expression. 
However, I want to avoid to write "don't provide the technical support" in the product documentation as much as possible. (As far as I know, the product doc says what the product can do and cannot do, and then this means we can support and cannot support.) 

I suppose: 
- **This may work** means we want to say it may work but it is not supported and tested. 
- The users may take it may work , should work without testing, but Scalar should support since the doc says it work.   (In the technical support world, the users always take positive to their side)

My suggestion is just describe what we support like  "We recommend using LTS versions mentioned above as the only supported JDK." or something like that. 
(We don't want to discuss about non-supported environment may/should work in the doc or not and if we describe it clearly, then this provide the negative support to the customer. FAQ or some other place can follow this up if possible, as it's true this may work technically. ) 

@josh-wong please advise me how to mention non-supported scenario in the doc?

Reviewer 6: @NatsukiMiyamoto  
OK. Let's discuss it in the support meeting! 🙇‍♀️
(I think we need to update both ScalarDB/ScalarDL documents.)

Reviewer: @NatsukiMiyamoto @kota2and3kan 
> @josh-wong please advise me how to mention non-supported scenario in the doc?

Below or a couple suggestions. I know these suggestions might seem a bit strongly worded, but I think we should be accurate rather than vague or unclear.

```suggestion
We recommend using the LTS versions mentioned above since they are the only JDK versions that we support.
```

Or maybe the following?:

```suggestion
We recommend using the LTS versions mentioned above since we have tested operability on only those JDK versions.
```

Reviewer 5: Thank you Josh. I prefer the first one. 
(I prefer to specify the supported version regardless tested or not.)

Reviewer: @NatsukiMiyamoto Totally agree👍

Reviewer 6: @josh-wong  
Thank you for your suggestion!
We decided to discuss it in the support meeting since we should apply this update to the ScalarDB docs side too.
So, in this PR, we don't apply this update.
We will do it in another PR after additional discussion.

@NatsukiMiyamoto  
As we mentioned in the support daily meeting, let's discuss it in the next support weekly meeting!

Author: @kota2and3kan @NatsukiMiyamoto So, I'm not applying the suggestion. Thanks!

In addition, other JDKs should work with ScalarDL, but we haven't tested them.

> {% endcapture %}
> 

Reviewer: This suggestion adds the closing string for the notice.

```suggestion
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

```

If you use JDK versions other than 8, specify `-release 8` option for `javac` or define what toolchain to use in your Gradle and Maven configurations.
For the Gradle and Maven configurations, check [the Gradle doc](https://docs.gradle.org/current/userguide/toolchains.html) and [the Maven doc](https://maven.apache.org/guides/mini/guide-using-toolchains.html).

{% capture notice--warning %}
**Attention**
Since ScalarDL is built with JDK 8, contracts must be a JDK 8-compatible binary.
If you use a version other than JDK 8, you must configure your build tool to build the JDK 8-compatible binary.
There are several ways to specify binary compatibility, including using the `--release 8` option for javac or setting Gradle (or Maven) configurations to use the JDK 8 toolchain as follows.

```gradle
java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(8))
    }
}
```

> For more details about thee Gradle and Maven configurations, see [Toolchains for JVM projects for Gradle](https://docs.gradle.org/current/userguide/toolchains.html) and [Guide to Using Toolchains for Maven](https://maven.apache.org/guides/mini/guide-using-toolchains.html).

Reviewer: ```suggestion
For more details about the Gradle and Maven configurations, see [Toolchains for JVM projects for Gradle](https://docs.gradle.org/current/userguide/toolchains.html) and [Guide to Using Toolchains for Maven](https://maven.apache.org/guides/mini/guide-using-toolchains.html).
```

> {% endcapture %}
> 

Reviewer: The following is necessary to style the Attention notice for the docs site.

```suggestion
{% endcapture %}

<div class="notice--warning">{{ notice--warning | markdownify }}</div>
```
